# 2. Selección de variables.

## Descripción

En este notebook se pretende crear una red neuronal empleando todas las definiciones de despoblamiento como target, uniéndolas en un único target. Además, se eliminarán variables que tengan una alta correlación con las variables objetivo.

## Librerías

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Carga de los datasets:

Norte de España:

In [2]:
excel_pob = pd.read_csv('/home/jovyan//PRÁCTICAS DESPOBLAMIENTO/DATOS/NUEVOS DATOS/excel_pob.csv', sep=',',encoding="utf-8")
excel_pob

/home/jovyan/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (98,99,100,101,102,103,119,120,121,122,123,124,140,141,142,143,144,145) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


CODMUNI  OBJECTID              NombreCCAA Provincia  \
0       33001         1  Principado de Asturias  Asturias   
1       33002         2  Principado de Asturias  Asturias   
2       33003         3  Principado de Asturias  Asturias   
3       33006         5  Principado de Asturias  Asturias   
4       33007         6  Principado de Asturias  Asturias   
...       ...       ...                     ...       ...   
3684    50296      3922                  Aragón  Zaragoza   
3685    50292      3923                  Aragón  Zaragoza   
3686    50295      3924                  Aragón  Zaragoza   
3687    50298      3926                  Aragón  Zaragoza   
3688    50293      3927                  Aragón  Zaragoza   

                    Municipio  MINMDT  MAXMDT  RANGEMDT      MEANMDT  \
0                     Allande     117    1394      1277   784.735210   
1                       Aller     271    2105      1834  1067.056635   
2                      Amieva      99    2393      2294   817.504449   
3                     Bimenes     274    1113       839   506.502204   
4                        Boal      33    1151      1118   493.553497   
...                       ...     ...     ...       ...          ...   
3684               Zaida (La)     139     300       161   189.077045   
3685     Villarreal de Huerva     831    1205       374   931.972503   
3686               Vistabella     722    1129       407   863.795376   
3687                    Zuera     241     770       529   413.819021   
3688  Villarroya de la Sierra     671    1408       737   896.568072   

          STDMDT  ...     DISTCURGH   DISTATPRIM  VIVTUR21  PLA21   KM2BLAN  \
0     233.924492  ...   5997.693739  5274.955751         2     13  0.001683   
1     382.986501  ...  11748.784830  3961.940782        58    264  0.026321   
2     455.771315  ...   8966.275234  4285.074165        21    112  0.000740   
3     142.357363  ...   5417.955543  2496.420733         9     51  0.005534   
4     218.406880  ...   4317.579792  4305.153512        15     72  0.018779   
...          ...  ...           ...          ...       ...    ...       ...   
3684   36.271257  ...   7899.678143  2227.740226         0      0  0.000000   
3685   73.191988  ...  13996.062990  2585.646631         0      0  0.000000   
3686   71.596410  ...   6210.190730  1943.478095         0      0  0.002940   
3687  110.728010  ...   8860.678727  5572.568876         1      2  0.415539   
3688  140.151654  ...   4471.936801  3520.124977         0      0  0.000139   

      KM2GRIS  CBMAS30  CBMAS100  COB3G  COB4G  
0           0     0.00      0.00   0.96   0.94  
1           0     1.00      0.00   1.00   1.00  
2           0     0.30      0.00   0.95   0.69  
3           0     0.17      0.00   1.00   1.00  
4           0     0.75      0.00   1.00   1.00  
...       ...      ...       ...    ...    ...  
3684        0     1.00      0.86   1.00   1.00  
3685        0     0.95      0.00   1.00   1.00  
3686        0     0.95      0.00   0.98   1.00  
3687        0     0.95      0.02   1.00   1.00  
3688        0     0.86      0.86   1.00   1.00  

[3689 rows x 586 columns]

Cantabria:

In [3]:
excel_cantabria = pd.read_csv('/home/jovyan//PRÁCTICAS DESPOBLAMIENTO/DATOS/NUEVOS DATOS/excel_cantabria.csv', sep=',')
excel_cantabria

CODMUNI  OBJECTID NombreCCAA  Provincia              Municipio  MINMDT  \
0     39084      2377  Cantabria  Cantabria              Solórzano      65   
1     39085      2378  Cantabria  Cantabria                Suances      -6   
2     39001      2379  Cantabria  Cantabria       Alfoz de Lloredo      -3   
3     39003      2380  Cantabria  Cantabria                Anievas     203   
4     39100      2381  Cantabria  Cantabria             Villafufre     112   
..      ...       ...        ...        ...                    ...     ...   
95    39097      2474  Cantabria  Cantabria            Vega de Pas     247   
96    39098      2475  Cantabria  Cantabria          Villacarriedo     150   
97    39099      2476  Cantabria  Cantabria            Villaescusa       0   
98    39101      2477  Cantabria  Cantabria  Villaverde de Trucíos     152   
99    39102      2478  Cantabria  Cantabria                   Voto      -3   

    MAXMDT  RANGEMDT     MEANMDT      STDMDT  ...    DISTCURGH   DISTATPRIM  \
0      552       487  220.690553  104.095285  ...  8655.201148  1503.589655   
1      218       224   57.649513   42.324675  ...  2249.691385  1682.295735   
2      391       394  141.156113   83.474135  ...  7292.339080  1657.810817   
3      942       739  467.784320  140.755881  ...  6239.139244  1794.658987   
4      828       716  329.966152  142.751930  ...  6234.089421  2070.778547   
..     ...       ...         ...         ...  ...          ...          ...   
95    1682      1435  716.320500  251.176248  ...  9058.263289  4048.590984   
96     870       720  429.996782  182.276557  ...  3363.345133  2203.640557   
97     437       437  112.661945   80.694858  ...  3405.989670  1269.354330   
98     735       583  363.969076  136.276553  ...  6938.838735  2412.441843   
99     808       811  217.173197  167.650140  ...  6154.359040  2932.312878   

    VIVTUR21  PLA21   KM2BLAN  KM2GRIS  CBMAS30  CBMAS100  COB3G  COB4G  
0          9     79  0.020988        0     0.60      0.00   0.98   0.98  
1        156    814  0.000103        0     0.99      0.62   1.00   1.00  
2         64    322  0.004960        0     0.96      0.00   1.00   1.00  
3          5     33  0.025491        0     0.00      0.00   1.00   1.00  
4          3     10  0.024887        0     0.98      0.02   1.00   1.00  
..       ...    ...       ...      ...      ...       ...    ...    ...  
95        10     50  0.183440        0     0.29      0.00   1.00   1.00  
96        23    169  0.058897        0     1.00      1.00   1.00   1.00  
97        10     64  0.004778        0     0.98      0.79   1.00   1.00  
98         0      0  0.000556        0     0.49      0.00   1.00   1.00  
99        35    197  0.912579        0     1.00      0.00   1.00   1.00  

[100 rows x 586 columns]

Se obtienen los targets siguiendo la definición de despoblamiento explicada en los notebooks anteriores.

In [4]:
def criterio_poblacion(y): #Se crea una pequeña función, puesto que hay que aplicarla en varias ocasiones.
    y[y <= 2000.0] = 1
    y[y > 2000.0] = 0
    return y

In [5]:
def criterio_densidad(y):
    y[y <= 12.50000] = 1
    y[y > 12.50000] = 0
    return y

In [6]:
def criterio_envejecimiento(y):
    y[y <= 30.0] = 0
    y[y > 30.0] = 1
    return y

Se seleccionan los datos de población y densidad de población para los años 2020 y 2021, en ambos datasets.

In [7]:
#España:
y_pob11 = excel_pob["POB2020"]
y_den11 = excel_pob["DEN2020"]

#Cantabria:
y_pobc11 = excel_cantabria["POB2020"]
y_denc11 = excel_cantabria["DEN2020"]

In [8]:
#España:
y_pob22 = excel_pob["POB2021"]
y_den22 = excel_pob["DEN2021"]
# y_tenv22 = excel_pob["TENVEJ21"] #Aquí también se selecciona la variable tasa de envejecimiento. No está presente en el año 2020.

#Cantabria:
y_pobc22 = excel_cantabria["POB2021"]
y_denc22 = excel_cantabria["DEN2021"]
# y_tenvc22 = excel_cantabria["TENVEJ21"]

Aplicamos las funciones creadas anteriormente para transformar las columnas en 0 y 1 para el problema de clasificación.

In [9]:
y_pob1 = criterio_poblacion(y_pob11)
y_pob2 = criterio_poblacion(y_pob22)
y_pobc1 = criterio_poblacion(y_pobc11)
y_pobc2 = criterio_poblacion(y_pobc22)
y_den1 = criterio_densidad(y_den11)
y_den2 = criterio_densidad(y_den22)
y_denc1 = criterio_densidad(y_denc11)
y_denc2 = criterio_densidad(y_denc22)

/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Es necesario unir las columnas de población en una sola. Para lograr este propósito, se realiza un bucle:

In [10]:
y_pob = [] #Inicialización de lista vacía, donde se almacenarán los resultados.
for i,j in zip(range(len(y_pob1)), range(len(y_pob2))):
    if y_pob1[i] == y_pob2[j]: #Si el valor de ambos coincide, se añade ese valor.
        y_pob.append(y_pob2[j])
    else:
        y_pob.append(1) #Si por el contrario no coincide, se determina que está en riesgo de despoblamiento.

In [11]:
y_pob

[1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0

Repetimos la operación para la densidad de población:

In [12]:
y_den = []
for i,j in zip(range(len(y_den1)), range(len(y_den2))):
    if y_den1[i] == y_den2[j]:
        y_den.append(y_den2[j])
    else:
        y_den.append(1)
y_den

[1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0

Y para Cantabria:

In [13]:
y_pobc = []
for i,j in zip(range(len(y_pobc1)), range(len(y_pobc2))):
    if y_pobc1[i] == y_pobc2[j]:
        y_pobc.append(y_pobc2[j])
    else:
        y_pobc.append(1)
y_pobc

[1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0]

In [14]:
y_denc = []
for i,j in zip(range(len(y_denc1)), range(len(y_denc2))):
    if y_denc1[i] == y_denc2[j]:
        y_denc.append(y_denc2[j])
    else:
        y_denc.append(1)
y_denc

[0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0]

Se repite la misma operación para el criterio tasa de envejecimiento. Ahora, sólo se cuentan con valores para el año 2021.

In [15]:
#España:
y_tenv = excel_pob["TENVEJ21"]

#Cantabria:
y_tenvc = excel_cantabria["TENVEJ21"]

In [16]:
#España
y_tenv[y_tenv < 30.0] = 0
y_tenv[y_tenv >= 30.0] = 1
y_tenv = y_tenv.astype('int64')
y_tenv

/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0       1
1       1
2       1
3       1
4       1
       ..
3684    1
3685    0
3686    1
3687    0
3688    1
Name: TENVEJ21, Length: 3689, dtype: int64

In [17]:
#Cantabria
y_tenvc[y_tenvc < 30.0] = 0
y_tenvc[y_tenvc >= 30.0] = 1
y_tenv

/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/home/jovyan/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0       1
1       1
2       1
3       1
4       1
       ..
3684    1
3685    0
3686    1
3687    0
3688    1
Name: TENVEJ21, Length: 3689, dtype: int64

Ahora, se diseña una pequeña función con la que se determinará el porcentaje de municipios en riesgo de despoblamiento:

In [18]:
def count_target(y):
    count_0 = y.value_counts()[0]
    count_1 = y.value_counts()[1]
    print('El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es : ', (count_0/len(y)*100))
    print('El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es : ', (count_1/len(y)*100))

In [19]:
count_target(pd.DataFrame(y_pob))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    11.059908
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    88.940092
dtype: float64


In [20]:
count_target(pd.DataFrame(y_pobc))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    45.0
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    55.0
dtype: float64


In [21]:
count_target(pd.DataFrame(y_den))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    32.285172
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    67.714828
dtype: float64


In [22]:
count_target(pd.DataFrame(y_denc))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    69.0
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    31.0
dtype: float64


In [23]:
count_target(pd.DataFrame(y_tenv))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  TENVEJ21
0           38.465709
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  TENVEJ21
1           61.534291
dtype: float64


In [24]:
count_target(pd.DataFrame(y_tenvc))

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  TENVEJ21
0.0         73.0
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  TENVEJ21
1.0         27.0
dtype: float64


Una vez vistos los porcentajes por separado, se procede a unirlos en una única variable objetivo.

In [25]:
#España
y_todas2 = []
for i,j in zip(range(len(y_pob)), range(len(y_den))):
    if y_den[i] == y_pob[j]:
        y_todas2.append(y_pob[j])
    else:
        y_todas2.append(1)

In [26]:
y_todas2

[1.0,
 0.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1,
 1,
 0.0,
 0.0,
 1,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 1,
 1.0,
 0.0,
 0.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1,
 1,
 1.0,
 0.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1,
 1,
 0.0,
 1,
 1,
 0.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 0.0,
 1.0,
 1.0,
 0.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1,
 1,
 1,
 1,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1,
 1.0,
 1.0,
 1,
 1.0,
 1,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1

In [27]:
#España
y_todas = []
for i,j in zip(range(len(y_todas2)), range(len(y_tenv))):
    if y_tenv[i] == y_todas2[j]:
        y_todas.append(y_todas2[j])
    else:
        y_todas.append(1)

In [28]:
y_todas = pd.DataFrame(y_todas)

In [29]:
count_target(y_todas)

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    10.246679
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    89.753321
dtype: float64


In [30]:
#Cantabria
y_todas1 = []
for i,j in zip(range(len(y_pobc)), range(len(y_denc))):
    if y_denc[i] == y_pobc[j]:
        y_todas1.append(y_pobc[j])
    else:
        y_todas1.append(1)

In [31]:
#Cantabria
y_todasC = []
for i,j in zip(range(len(y_todas1)), range(len(y_tenvc))):
    if y_tenvc[i] == y_todas1[j]:
        y_todasC.append(y_todas1[j])
    else:
        y_todasC.append(1)

In [32]:
y_todasC = pd.DataFrame(y_todasC)

In [33]:
count_target(y_todasC)

El porcentaje de las muestras que se corresponde con municipios que están no en riesgo de despoblamiento es :  0.0    44.0
dtype: float64
El porcentaje de las muestras que se corresponde con municipios que están en riesgo de despoblamiento es :  1.0    56.0
dtype: float64


Transformamos todas las varables objetivo a categoricas:

In [34]:
y_pob = pd.DataFrame(y_pob)
y_pobc = pd.DataFrame(y_pobc)
y_den = pd.DataFrame(y_den)
y_denc = pd.DataFrame(y_denc)
y_tenv = pd.DataFrame(y_tenv)
y_tenvc = pd.DataFrame(y_tenvc)

In [35]:
y_pob = y_pob.astype("category")
y_pobc = y_pobc.astype("category")
y_den = y_den.astype("category")
y_denc = y_denc.astype("category")
y_tenv = y_tenv.astype("category")
y_tenvc = y_tenvc.astype("category")

In [36]:
y_todas = y_todas.astype("category")
y_todasC = y_todasC.astype("category")

Eliminamos las variables ObjectID, Provincia y MUNICIPIO ya que no aportan nada a la red neuronal. Más adelante se eliminará también la variable CODMUNI y NombreCCAA, pero de momento la dejamos por si hace falta unir algún csv a mayores.

In [37]:
excel_pob = excel_pob.drop(['OBJECTID'], axis=1)
excel_pob = excel_pob.drop(['Provincia'], axis=1)
excel_pob = excel_pob.drop(['Municipio'], axis=1)

excel_cantabria = excel_cantabria.drop(['OBJECTID'], axis=1)
excel_cantabria = excel_cantabria.drop(['Provincia'], axis=1)
excel_cantabria = excel_cantabria.drop(['Municipio'], axis=1)

## Despoblamiento en el Norte de España:

El despoblamiento en el norte de España tiene un comportamiento diferente al del sur de España. En concreto, las comunimadades autónomas que componen el despoblamiento en esta zona son:
* Aragón
* Asturias
* Cantabria
* Castilla y León
* La Rioja
* Navarra
* Pais Vasco

In [38]:
excel_pob['NombreCCAA'].value_counts()

Castilla y León               2185
Aragón                         690
Comunidad Foral de Navarra     260
País Vasco                     210
La Rioja                       170
Cantabria                      100
Principado de Asturias          74
Name: NombreCCAA, dtype: int64

Como podemos comprobar, en el dataset sólo hay municipios de esta zona.

## Variables a eliminar:

Algunas de las variables se van a eliminar por ser redundantes a la hora de la realización de la red neuronal. Estas variables son las siguientes:

    - Tasa de crecimiento medio interanual (Variables que contienen 'TC', aquí también se incluyen las tasas de crecimiento de hombres y mujeres)
    - Población masculina (Variables que comienzan por 'HOM')
    - Población femenina (Variables que comienzan por 'MUJ')
    - Saldo vegetativo (Variables que comienzan por 'SVEG'): Por ser la diferencia entre la natalidad y la mortalidad, no aporta nada nuevo.
    - Tasa bruta de natalidad (Variables que comienzan por 'TBNAT')
    - Tasa bruta de mortalidad (Variables que comienzan por 'TBMOR')
    - Tasa crecimiento vegetativo (Variables que comienzan por 'TCVEG')
    - Todas las variables que contienen población de un cierto rango de edad (0A4 años, 5A9 años...) tanto a nivel global como a nivel de hombres y mujeres).
    - Todas las variables que contienen JOV, VIE, SO o ADUL por contener varibles desglosadas por la población por edades.
    - Porcentaje de población adulta ((Variables que comienzan por 'PRPAD')
    - Todas las variables relacionadas con población, densidad de población y tasa de envejecimiento, puesto que están en la definición de despoblamiento.
    - Mujeres en edad de fecunda (Variables que comienzan por 'MEFE')


 Las tasas se han eliminado por tener relación directa con la variable población o con otras variables como la natalidad o la mortalidad. Sucede lo mismo con los porcentajes.

In [39]:
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('TC')] #Elimina las columnas que comienzan por 'TC'.
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('HOM')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('MUJ')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('TBNAT')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('TBMOR')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('TCVEG')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('SVEG')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('0A4')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('5A9')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('10A14')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('15A19')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('20A24')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('25A29')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('30A34')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('35A39')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('40A44')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('45A49')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('50A54')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('55A59')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('60A64')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('65A69')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('70A74')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('75A79')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('80A84')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('85A89')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('90A94')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('95A99')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('100')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('JOV')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('VIE')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('ADUL')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('SO')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('PRPAD')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('MEFE')]
excel_pob = excel_pob.drop(['KM2GRIS'], axis=1)
excel_pob = excel_pob.drop(['KM2BLAN'], axis=1)

# Estas variables se han eliminado tras el estudio de correlaciones visto en el notebook 3, por su alta correlación.
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('DEPEN')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('NAT')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('MOR')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('BAJAS')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('ALTAS')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('IDV')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('SALDO')]

In [40]:
excel_pob

CODMUNI              NombreCCAA  MINMDT  MAXMDT  RANGEMDT      MEANMDT  \
0       33001  Principado de Asturias     117    1394      1277   784.735210   
1       33002  Principado de Asturias     271    2105      1834  1067.056635   
2       33003  Principado de Asturias      99    2393      2294   817.504449   
3       33006  Principado de Asturias     274    1113       839   506.502204   
4       33007  Principado de Asturias      33    1151      1118   493.553497   
...       ...                     ...     ...     ...       ...          ...   
3684    50296                  Aragón     139     300       161   189.077045   
3685    50292                  Aragón     831    1205       374   931.972503   
3686    50295                  Aragón     722    1129       407   863.795376   
3687    50298                  Aragón     241     770       529   413.819021   
3688    50293                  Aragón     671    1408       737   896.568072   

          STDMDT  MINPENDI   MAXPENDI  RANGEPENDI  ...  NFARMA  NCENTEDUC  \
0     233.924492  0.173360  46.501644   46.328284  ...       1          2   
1     382.986501  0.122584  61.203892   61.081307  ...       4         10   
2     455.771315  0.346717  67.186058   66.839341  ...       0          0   
3     142.357363  0.122584  43.220619   43.098035  ...       1          8   
4     218.406880  0.000000  42.763451   42.763451  ...       2          0   
...          ...       ...        ...         ...  ...     ...        ...   
3684   36.271257  0.000000  13.273371   13.273371  ...       0          0   
3685   73.191988  0.000000  26.876348   26.876348  ...       0          0   
3686   71.596410  0.173360  34.187431   34.014071  ...       0          0   
3687  110.728010  0.000000  27.790625   27.790625  ...       1         10   
3688  140.151654  0.000000  32.033096   32.033096  ...       0          0   

       DISTHOSPIT     DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  \
0     16583.21167   5709.556279  5274.955751         2     13     0.00   0.96   
1     19974.83089   4971.806926  3961.940782        58    264     1.00   1.00   
2     17540.29587   5485.216677  4285.074165        21    112     0.30   0.95   
3     10146.40817   6522.937789  2496.420733         9     51     0.17   1.00   
4     15694.48914   3757.152641  4305.153512        15     72     0.75   1.00   
...           ...           ...          ...       ...    ...      ...    ...   
3684  41271.99526   8612.545190  2227.740226         0      0     1.00   1.00   
3685  35404.93621  18222.786360  2585.646631         0      0     0.95   1.00   
3686  43033.52273  10831.081460  1943.478095         0      0     0.95   0.98   
3687  22120.23364   5762.972359  5572.568876         1      2     0.95   1.00   
3688  20872.94112  13550.215130  3520.124977         0      0     0.86   1.00   

      COB4G  
0      0.94  
1      1.00  
2      0.69  
3      1.00  
4      1.00  
...     ...  
3684   1.00  
3685   1.00  
3686   1.00  
3687   1.00  
3688   1.00  

[3689 rows x 142 columns]

In [41]:
#Las variables que contienen estos años están relacionadas con la población. Además, no interesa tener variables tan lejanas 
# en el tiempo, puesto que las condiciones han podido cambiar mucho desde entonces.
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1900')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1910')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1920')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1930')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1940')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1950')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1960')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.contains('1970')]

In [42]:
excel_pob

CODMUNI              NombreCCAA  MINMDT  MAXMDT  RANGEMDT      MEANMDT  \
0       33001  Principado de Asturias     117    1394      1277   784.735210   
1       33002  Principado de Asturias     271    2105      1834  1067.056635   
2       33003  Principado de Asturias      99    2393      2294   817.504449   
3       33006  Principado de Asturias     274    1113       839   506.502204   
4       33007  Principado de Asturias      33    1151      1118   493.553497   
...       ...                     ...     ...     ...       ...          ...   
3684    50296                  Aragón     139     300       161   189.077045   
3685    50292                  Aragón     831    1205       374   931.972503   
3686    50295                  Aragón     722    1129       407   863.795376   
3687    50298                  Aragón     241     770       529   413.819021   
3688    50293                  Aragón     671    1408       737   896.568072   

          STDMDT  MINPENDI   MAXPENDI  RANGEPENDI  ...  NFARMA  NCENTEDUC  \
0     233.924492  0.173360  46.501644   46.328284  ...       1          2   
1     382.986501  0.122584  61.203892   61.081307  ...       4         10   
2     455.771315  0.346717  67.186058   66.839341  ...       0          0   
3     142.357363  0.122584  43.220619   43.098035  ...       1          8   
4     218.406880  0.000000  42.763451   42.763451  ...       2          0   
...          ...       ...        ...         ...  ...     ...        ...   
3684   36.271257  0.000000  13.273371   13.273371  ...       0          0   
3685   73.191988  0.000000  26.876348   26.876348  ...       0          0   
3686   71.596410  0.173360  34.187431   34.014071  ...       0          0   
3687  110.728010  0.000000  27.790625   27.790625  ...       1         10   
3688  140.151654  0.000000  32.033096   32.033096  ...       0          0   

       DISTHOSPIT     DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  \
0     16583.21167   5709.556279  5274.955751         2     13     0.00   0.96   
1     19974.83089   4971.806926  3961.940782        58    264     1.00   1.00   
2     17540.29587   5485.216677  4285.074165        21    112     0.30   0.95   
3     10146.40817   6522.937789  2496.420733         9     51     0.17   1.00   
4     15694.48914   3757.152641  4305.153512        15     72     0.75   1.00   
...           ...           ...          ...       ...    ...      ...    ...   
3684  41271.99526   8612.545190  2227.740226         0      0     1.00   1.00   
3685  35404.93621  18222.786360  2585.646631         0      0     0.95   1.00   
3686  43033.52273  10831.081460  1943.478095         0      0     0.95   0.98   
3687  22120.23364   5762.972359  5572.568876         1      2     0.95   1.00   
3688  20872.94112  13550.215130  3520.124977         0      0     0.86   1.00   

      COB4G  
0      0.94  
1      1.00  
2      0.69  
3      1.00  
4      1.00  
...     ...  
3684   1.00  
3685   1.00  
3686   1.00  
3687   1.00  
3688   1.00  

[3689 rows x 126 columns]

In [43]:
#Eliminamos las variables relacionadas con la variable objetivo.
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('POB')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('DEN')]
excel_pob = excel_pob.loc[:, ~excel_pob.columns.str.startswith('TENV')]

In [44]:
excel_pob

CODMUNI              NombreCCAA  MINMDT  MAXMDT  RANGEMDT      MEANMDT  \
0       33001  Principado de Asturias     117    1394      1277   784.735210   
1       33002  Principado de Asturias     271    2105      1834  1067.056635   
2       33003  Principado de Asturias      99    2393      2294   817.504449   
3       33006  Principado de Asturias     274    1113       839   506.502204   
4       33007  Principado de Asturias      33    1151      1118   493.553497   
...       ...                     ...     ...     ...       ...          ...   
3684    50296                  Aragón     139     300       161   189.077045   
3685    50292                  Aragón     831    1205       374   931.972503   
3686    50295                  Aragón     722    1129       407   863.795376   
3687    50298                  Aragón     241     770       529   413.819021   
3688    50293                  Aragón     671    1408       737   896.568072   

          STDMDT  MINPENDI   MAXPENDI  RANGEPENDI  ...  NFARMA  NCENTEDUC  \
0     233.924492  0.173360  46.501644   46.328284  ...       1          2   
1     382.986501  0.122584  61.203892   61.081307  ...       4         10   
2     455.771315  0.346717  67.186058   66.839341  ...       0          0   
3     142.357363  0.122584  43.220619   43.098035  ...       1          8   
4     218.406880  0.000000  42.763451   42.763451  ...       2          0   
...          ...       ...        ...         ...  ...     ...        ...   
3684   36.271257  0.000000  13.273371   13.273371  ...       0          0   
3685   73.191988  0.000000  26.876348   26.876348  ...       0          0   
3686   71.596410  0.173360  34.187431   34.014071  ...       0          0   
3687  110.728010  0.000000  27.790625   27.790625  ...       1         10   
3688  140.151654  0.000000  32.033096   32.033096  ...       0          0   

       DISTHOSPIT     DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  \
0     16583.21167   5709.556279  5274.955751         2     13     0.00   0.96   
1     19974.83089   4971.806926  3961.940782        58    264     1.00   1.00   
2     17540.29587   5485.216677  4285.074165        21    112     0.30   0.95   
3     10146.40817   6522.937789  2496.420733         9     51     0.17   1.00   
4     15694.48914   3757.152641  4305.153512        15     72     0.75   1.00   
...           ...           ...          ...       ...    ...      ...    ...   
3684  41271.99526   8612.545190  2227.740226         0      0     1.00   1.00   
3685  35404.93621  18222.786360  2585.646631         0      0     0.95   1.00   
3686  43033.52273  10831.081460  1943.478095         0      0     0.95   0.98   
3687  22120.23364   5762.972359  5572.568876         1      2     0.95   1.00   
3688  20872.94112  13550.215130  3520.124977         0      0     0.86   1.00   

      COB4G  
0      0.94  
1      1.00  
2      0.69  
3      1.00  
4      1.00  
...     ...  
3684   1.00  
3685   1.00  
3686   1.00  
3687   1.00  
3688   1.00  

[3689 rows x 67 columns]

Para no tener que repetir todo el proceso para Cantabria, filtramos en este nuevo dataset con menos variables:

In [45]:
excel_cantabria = excel_pob[excel_pob['NombreCCAA'] == 'Cantabria'] 
excel_pob = excel_pob.drop(['NombreCCAA'], axis=1) #Eliminamos la variable NombreCCAA ya que no nos va a hacer más falta.
excel_pob = excel_pob.drop(['CODMUNI'], axis=1)
excel_cantabria = excel_cantabria.drop(['CODMUNI'], axis=1) #Eliminamos también el código de municipio, puesto que no es una
# variable numérica

In [46]:
excel_cantabria

NombreCCAA  MINMDT  MAXMDT  RANGEMDT     MEANMDT      STDMDT  MINPENDI  \
2259  Cantabria      65     552       487  220.690553  104.095285  0.000000   
2260  Cantabria      -6     218       224   57.649513   42.324675  0.000000   
2261  Cantabria      -3     391       394  141.156113   83.474135  0.122584   
2262  Cantabria     203     942       739  467.784320  140.755881  0.173360   
2263  Cantabria     112     828       716  329.966152  142.751930  0.122584   
...         ...     ...     ...       ...         ...         ...       ...   
2354  Cantabria     247    1682      1435  716.320500  251.176248  0.122584   
2355  Cantabria     150     870       720  429.996782  182.276557  0.000000   
2356  Cantabria       0     437       437  112.661945   80.694858  0.122584   
2357  Cantabria     152     735       583  363.969076  136.276553  0.274105   
2358  Cantabria      -3     808       811  217.173197  167.650140  0.000000   

       MAXPENDI  RANGEPENDI  MEANPENDI  ...  NFARMA  NCENTEDUC    DISTHOSPIT  \
2259  31.196602   31.196602  11.105984  ...       1          0  13305.256790   
2260  23.273787   23.273787   5.522714  ...       2          4   6086.627482   
2261  33.228851   33.106267   8.968757  ...       0          2   9033.361309   
2262  37.476574   37.303214  14.057126  ...       0          0  18892.149520   
2263  36.627811   36.505227  12.574647  ...       0          0  18897.890150   
...         ...         ...        ...  ...     ...        ...           ...   
2354  62.624386   62.501801  22.598704  ...       1          2  33837.993620   
2355  39.624466   39.624466  12.713040  ...       1          2  24542.374170   
2356  27.935255   27.812671   8.587506  ...       0          0   9800.767314   
2357  29.921757   29.647651  10.768345  ...       0          0  22527.063100   
2358  41.322651   41.322651  12.887623  ...       1          2   9344.076767   

        DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  COB4G  
2259  6689.805575  1503.589655         9     79     0.60   0.98   0.98  
2260  3558.752869  1682.295735       156    814     0.99   1.00   1.00  
2261  4085.905018  1657.810817        64    322     0.96   1.00   1.00  
2262  9992.467164  1794.658987         5     33     0.00   1.00   1.00  
2263  5427.973068  2070.778547         3     10     0.98   1.00   1.00  
...           ...          ...       ...    ...      ...    ...    ...  
2354  5052.512126  4048.590984        10     50     0.29   1.00   1.00  
2355  3364.998813  2203.640557        23    169     1.00   1.00   1.00  
2356  9562.771833  1269.354330        10     64     0.98   1.00   1.00  
2357  7025.784394  2412.441843         0      0     0.49   1.00   1.00  
2358  3885.421461  2932.312878        35    197     1.00   1.00   1.00  

[100 rows x 66 columns]

In [47]:
excel_cantabria = excel_cantabria.drop(['NombreCCAA'], axis=1)

In [48]:
excel_pob

MINMDT  MAXMDT  RANGEMDT      MEANMDT      STDMDT  MINPENDI   MAXPENDI  \
0        117    1394      1277   784.735210  233.924492  0.173360  46.501644   
1        271    2105      1834  1067.056635  382.986501  0.122584  61.203892   
2         99    2393      2294   817.504449  455.771315  0.346717  67.186058   
3        274    1113       839   506.502204  142.357363  0.122584  43.220619   
4         33    1151      1118   493.553497  218.406880  0.000000  42.763451   
...      ...     ...       ...          ...         ...       ...        ...   
3684     139     300       161   189.077045   36.271257  0.000000  13.273371   
3685     831    1205       374   931.972503   73.191988  0.000000  26.876348   
3686     722    1129       407   863.795376   71.596410  0.173360  34.187431   
3687     241     770       529   413.819021  110.728010  0.000000  27.790625   
3688     671    1408       737   896.568072  140.151654  0.000000  32.033096   

      RANGEPENDI  MEANPENDI   STDPENDI  ...  NFARMA  NCENTEDUC   DISTHOSPIT  \
0      46.328284  20.298575   8.513073  ...       1          2  16583.21167   
1      61.081307  22.556711   8.204898  ...       4         10  19974.83089   
2      66.839341  25.832286  11.384265  ...       0          0  17540.29587   
3      43.098035  15.395220   6.130982  ...       1          8  10146.40817   
4      42.763451  17.966830   7.593164  ...       2          0  15694.48914   
...          ...        ...        ...  ...     ...        ...          ...   
3684   13.273371   2.960418   2.437278  ...       0          0  41271.99526   
3685   26.876348   7.047567   5.670949  ...       0          0  35404.93621   
3686   34.014071  10.970295   5.792647  ...       0          0  43033.52273   
3687   27.790625   3.889147   3.862538  ...       1         10  22120.23364   
3688   32.033096   6.590386   4.319270  ...       0          0  20872.94112   

         DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  COB4G  
0      5709.556279  5274.955751         2     13     0.00   0.96   0.94  
1      4971.806926  3961.940782        58    264     1.00   1.00   1.00  
2      5485.216677  4285.074165        21    112     0.30   0.95   0.69  
3      6522.937789  2496.420733         9     51     0.17   1.00   1.00  
4      3757.152641  4305.153512        15     72     0.75   1.00   1.00  
...            ...          ...       ...    ...      ...    ...    ...  
3684   8612.545190  2227.740226         0      0     1.00   1.00   1.00  
3685  18222.786360  2585.646631         0      0     0.95   1.00   1.00  
3686  10831.081460  1943.478095         0      0     0.95   0.98   1.00  
3687   5762.972359  5572.568876         1      2     0.95   1.00   1.00  
3688  13550.215130  3520.124977         0      0     0.86   1.00   1.00  

[3689 rows x 65 columns]

Veamos las variables restantes:

In [49]:
for col_name in excel_pob.reindex(sorted(excel_pob.columns), axis=1): 
    print(col_name)

AREAKM2
CARMAT03
CARMAT10
CARMAT21
CBMAS30
COB3G
COB4G
DISTATPRIM
DISTFARMA
DISTHOSPIT
HIJMUJ03
HIJMUJ10
HIJMUJ21
IENVEJ03
IENVEJ10
IENVEJ21
IREEMP03
IREEMP10
IREEMP21
MAXCARRETN
MAXDISAUT
MAXDISFERR
MAXDISN10M
MAXMDT
MAXPENDI
MEANCARRETN
MEANDISAUT
MEANDISFERR
MEANDISN10M
MEANMDT
MEANPENDI
MINCARRETN
MINDISAUT
MINDISFERR
MINDISN10M
MINMDT
MINPENDI
NCENTEDUC
NFARMA
NHOSPI
PLA21
PRMEF03
PRMEF10
PRMEF21
RANGECARRETN
RANGEDISAUT
RANGEDISFERR
RANGEDISN10M
RANGEMDT
RANGEPENDI
RMASCU03
RMASCU10
RMASCU21
STDCARRETN
STDDISAUT
STDDISFERR
STDDISN10M
STDMDT
STDPENDI
Shape_Area
Shape_Length
TJUVEN03
TJUVEN10
TJUVEN21
VIVTUR21


In [50]:
excel_cantabria

MINMDT  MAXMDT  RANGEMDT     MEANMDT      STDMDT  MINPENDI   MAXPENDI  \
2259      65     552       487  220.690553  104.095285  0.000000  31.196602   
2260      -6     218       224   57.649513   42.324675  0.000000  23.273787   
2261      -3     391       394  141.156113   83.474135  0.122584  33.228851   
2262     203     942       739  467.784320  140.755881  0.173360  37.476574   
2263     112     828       716  329.966152  142.751930  0.122584  36.627811   
...      ...     ...       ...         ...         ...       ...        ...   
2354     247    1682      1435  716.320500  251.176248  0.122584  62.624386   
2355     150     870       720  429.996782  182.276557  0.000000  39.624466   
2356       0     437       437  112.661945   80.694858  0.122584  27.935255   
2357     152     735       583  363.969076  136.276553  0.274105  29.921757   
2358      -3     808       811  217.173197  167.650140  0.000000  41.322651   

      RANGEPENDI  MEANPENDI  STDPENDI  ...  NFARMA  NCENTEDUC    DISTHOSPIT  \
2259   31.196602  11.105984  6.565275  ...       1          0  13305.256790   
2260   23.273787   5.522714  4.048844  ...       2          4   6086.627482   
2261   33.106267   8.968757  5.108402  ...       0          2   9033.361309   
2262   37.303214  14.057126  5.779102  ...       0          0  18892.149520   
2263   36.505227  12.574647  6.543079  ...       0          0  18897.890150   
...          ...        ...       ...  ...     ...        ...           ...   
2354   62.501801  22.598704  8.880146  ...       1          2  33837.993620   
2355   39.624466  12.713040  6.913056  ...       1          2  24542.374170   
2356   27.812671   8.587506  5.689353  ...       0          0   9800.767314   
2357   29.647651  10.768345  5.145302  ...       0          0  22527.063100   
2358   41.322651  12.887623  7.796029  ...       1          2   9344.076767   

        DISTFARMA   DISTATPRIM  VIVTUR21  PLA21  CBMAS30  COB3G  COB4G  
2259  6689.805575  1503.589655         9     79     0.60   0.98   0.98  
2260  3558.752869  1682.295735       156    814     0.99   1.00   1.00  
2261  4085.905018  1657.810817        64    322     0.96   1.00   1.00  
2262  9992.467164  1794.658987         5     33     0.00   1.00   1.00  
2263  5427.973068  2070.778547         3     10     0.98   1.00   1.00  
...           ...          ...       ...    ...      ...    ...    ...  
2354  5052.512126  4048.590984        10     50     0.29   1.00   1.00  
2355  3364.998813  2203.640557        23    169     1.00   1.00   1.00  
2356  9562.771833  1269.354330        10     64     0.98   1.00   1.00  
2357  7025.784394  2412.441843         0      0     0.49   1.00   1.00  
2358  3885.421461  2932.312878        35    197     1.00   1.00   1.00  

[100 rows x 65 columns]

Una vez eliminadas estas variables, podemos ver que nos hemos quedado con 65 variables de las 636 que teníamos en un principio. Veamos a continuación algunas de las variables que se han decidido dejar en la red neuronal:

    - Todas las variables contenidas en los excels 110, 111 y 112. Estos contienen datos geográficos, tales como altitud, latitud, distancia a carreteras o a ferrocarriles, etc.
    - Natalidad (Variables que comienzan por 'NAT')
    - Mortalidad (Variables que comienzan por 'MOR')
    - Altas totales por municipio (Variables que comienzan por 'ALTASTT')
    - Bajas totales por municipio (Variables que comienzan por 'BAJASTT')
    - Altas interiores por municipio (Variables que comienzan por 'ALTASINT')
    - Bajas interiores por municipio (Variables que comienzan por 'BAJASINT')
    - Altas exteriores por municipio (Variables que comienzan por 'ALTASEXT')
    - Bajas exteriores por municipio (Variables que comienzan por 'BAJASEXT')
    - Saldo migratorio total por municipio (Variables que comienzan por 'SALDOTT')
    - Saldo migratorio interior por municipio (Variables que comienzan por 'SALDOINT')
    - Saldo migratorio exterior por municipio (Variables que comienzan por 'SALDOEXT')
    - Número de hijos (de 0 a 4 años) por cada 100 mujeres (Variables que comienzan por 'HIJMUJ')
    - Índice de carga materna: niñas de 0 a 4 años por cada 100 mujeres en edad genésica (Variables que comienzan por 'CARMAT')

## Guardado de Datasets

Guardamos los nuevos datasets.

In [51]:
excel_pob.to_csv('excel_pob_variables.csv', index=False)
excel_cantabria.to_csv('excel_cantabria_variables.csv', index=False)

Guardamos el target para futuros análisis:

In [52]:
y_todas.to_csv('y_esp.csv', index=False)
y_todasC.to_csv('y_can.csv', index=False)